### PR TITLE
Add smoothing parameter to Discrete Cosine transform and reparam

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2426 

This adds a `smooth` parameter to the `DiscreteCosineTransform` and `DiscreteCosineReparam` allowing the reparameterizer to automatically encode C-smooth functions. That is, `DiscreteCosineReparam(smooth=s)` transforms C-s functions to white noise, e.g. `DCR(smooth=1)` transforms brownian motion to white noise, and `DCR(smooth=2)` transforms cumsum(cumsum(white noise)) to white noise.

I am hoping this will be useful in SVI and HMC where by setting `smooth=1` e.g. we would avoid the need for `AutoNormal` or `HMC` adaptation to learn the frequency-proportional variance (or rather we would warm start those learned variances at a better place).

## Tested
- added unit test cases for transform and reparam

### CODE REVIEW

File: pyro/distributions/transforms/discrete_cosine.py [hunk 1]
```diff
@@ -16,24 +17,45 @@ class DiscreteCosineTransform(Transform):
...
-        return type(self) == type(other) and self.event_dim == other.event_dim
+        return (type(self) == type(other) and self.event_dim == other.event_dim
+                and self.smooth == other.smooth)
+
+    @torch.no_grad()
+    def _weight(self, y):
+        size = y.size(-1)
+        if self._weight_cache is None or self._weight_cache.size(-1) != size:
+            # Weight by frequency**smooth, where the DCT-II frequencies are:
+            freq = torch.arange(0.5, size - 0.5, size, dtype=y.dtype, device=y.device)
```
Comment (fehiepsi) [inline pyro/distributions/transforms/discrete_cosine.py:46]: typo: `torch.arange(0.5, size - 0.5, dtype=...)`

File: pyro/distributions/transforms/discrete_cosine.py [hunk 2]
```diff
@@ -16,24 +17,45 @@ class DiscreteCosineTransform(Transform):
...
+            w = freq.pow_(self.smooth)
+            w /= w.log().mean().exp()  # Ensure |jacobian| = 1.
```
Comment (fehiepsi) [inline pyro/distributions/transforms/discrete_cosine.py:48]: nice trick to prevent blow up!

File: pyro/distributions/transforms/discrete_cosine.py [hunk 3]
```diff
@@ -16,24 +17,45 @@ class DiscreteCosineTransform(Transform):
...
+            self._weight_cache = w
+        return self._weight_cache
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply pyro/distributions/transforms/discrete_cosine.py:50]: I checked `weight by frequency**smooth` (up to constant multiplication - which does not matter after normalization) and `|jacobian|=1`. So this looks correct to me. However, I don't understand the phrase
> When 0, this transforms white noise to white noise;
when 1 this transforms continuous brownian-like motion to white noise;
when 2 this transforms doubly-cumsummed white

Could you elaborate it (maybe give some reference)?
        noise to white noise;

[the remark this answers is not in the document]
Comment (fritzo) [reply pyro/distributions/transforms/discrete_cosine.py]: The terminology comes from https://en.wikipedia.org/wiki/Colors_of_noise . I have also added a blurb to the `DiscreteCosineReparam` docstring.

The intuition is that we are treating this kind of like a linear normalizing flow where the codomain distribution should be approximately iid standard normal aka white noise. By setting `smooth=1` the domain distribution is brownian motion (i.e. cumsum or integral of white noise). By setting `smooth=2` the domain distribution is doubly-integrated white noise which is continuously differentiable.

Another piece of intuition is that this is a discrete analog of the Matern kernels for continuous GPs: smooth=0 is like Matern 1/2, smooth=1 is like Matern 3/2.

[the remark this answers is not in the document]
Comment (fehiepsi) [reply pyro/distributions/transforms/discrete_cosine.py]: Thanks, I made some plots to test your claims. It looks correct!
